apps/src/templates/FixZoomHelper.jsx

```diff
@@ -0,0 +1,183 @@
+    transform: 'translate(-50%,-50%)',
+    color: color.white,
+    backgroundColor: color.black,
+    opacity: 0.8,
+    padding: 15,
+    borderRadius: 5,
+    textAlign: 'center',
+    pointerEvents: 'none'
+  },
+  helperInnerIcons: {
+    fontSize: 24,
+    marginBottom: 10
+  },
+  closeX: {
+    position: 'absolute',
+    top: 5,
+    right: 5,
+    fontSize: 13
+  }
+};
+
+/**
+ * If the browser supports window.visualViewport, and we detect that the user
+ * is zoomed in, first show a small button in the bottom-right of the viewport.
+ * If pressed, a semi-transparent overlay is shown, which will also send touch
+ * events to the browser itself so that the user can pinch-to-zoom back to
+ * normal.  If the user fully zooms out, or taps once to dismiss, then the
+ * overlay will be hidden again.
+ */
+export default class FixZoomHelper extends React.Component {
+  // mode can be "none", "button", or "helper".
+  state = {
+    mode: 'none'
+  };
+
+  isLandscape() {
+    return window.matchMedia('(orientation: landscape)').matches;
+  }
+
+  isZoomed() {
+    return (
+      Math.abs(document.body.offsetWidth - window.visualViewport.width) > 30
+    );
+  }
+
+  updateViewport = () => {
+    const isZoomed = this.isZoomed();
+
+    if (this.state.mode !== 'none' && !isZoomed) {
+      const lastMode = this.state.mode;
+
+      // If transitioning from zoomed to non-zoomed, then mode goes to "none".
+      this.setState({mode: 'none'});
+
+      trackEvent('Research', 'FixZoomHelper', `${lastMode}-to-none`);
+    } else if (this.state.mode === 'none' && isZoomed) {
+      // If transitioning from non-zoomed to zoomed, then mode goes to "button".
+      this.setState({mode: 'button'});
+
+      trackEvent('Research', 'FixZoomHelper', 'none-to-button');
+    }
+
+    // Also update the viewport information.
+    this.setState({
+      offsetLeft: window.visualViewport.offsetLeft,
+      offsetTop: window.visualViewport.offsetTop,
+      width: window.visualViewport.width,
+      height: window.visualViewport.height
+    });
+  };
+
+  componentDidMount() {
+    // This component does nothing without visual viewport support.
+    if (window.visualViewport) {
+      this.updateViewport();
+
+      this.resizeListener = _.throttle(this.updateViewport, 200);
+      window.visualViewport.addEventListener('resize', this.resizeListener);
+      window.visualViewport.addEventListener('scroll', this.resizeListener);
+    }
+  }
+
+  componentWillUnmount() {
+    if (window.visualViewport) {
+      document.removeEventListener('resize', this.resizeListener, false);
+      document.removeEventListener('scroll', this.resizeListener, false);
+    }
+  }
+
+  onButtonClick = () => {
+    this.setState({mode: 'helper'});
+
+    trackEvent('Research', 'FixZoomHelper', 'button-to-helper');
+  };
+
+  onHelperClick = () => {
+    this.setState({mode: 'button'});
+
+    trackEvent('Research', 'FixZoomHelper', 'helper-to-button');
+  };
+
+  render() {
```

Comment:
Since you're accessing several state variables, consider migrating to this format:
`const {offsetTop, offsetLeft, height, width, mode} = this.state`